File: apps/web/src/types/transaction.ts
export type TransactionType = "income" | "expense";

export const TRANSACTION_RECURRENCE_TYPES = ["monthly"] as const;

export type TransactionRecurrenceType =
  (typeof TRANSACTION_RECURRENCE_TYPES)[number];

export const TRANSACTION_RECURRENCE_LABELS: Record<
  TransactionRecurrenceType,
  string
> = {
  monthly: "Mensal",
};

export const TRANSACTION_KINDS = [
  "single",
  "installment-template",
  "installment-instance",
  "recurring-template",
  "recurring-instance",
] as const;

export type TransactionKind = (typeof TRANSACTION_KINDS)[number];

export const TRANSACTION_RECURRENCE_MODES = [
  "indefinite",
  "until-date",
  "for-months",
] as const;

export type TransactionRecurrenceMode =
  (typeof TRANSACTION_RECURRENCE_MODES)[number];

export type Transaction = {
  id: string;
  title: string;
  amount: number;
  type: TransactionType;
  category: string;
  transactionKind: TransactionKind;
  sourceId: string | null;
  occurrenceDate: string | null;
  installmentIndex: number | null;
  installmentCount: number | null;
  installmentStartDate: string | null;
  recurringSourceId: string | null;
  recurringOccurrenceDate: string | null;
  isRecurring: boolean;
  recurrenceType: TransactionRecurrenceType | null;
  recurrenceMode: TransactionRecurrenceMode | null;
  recurrenceDay: number | null;
  recurrenceStartDate: string | null;
  recurrenceEndDate: string | null;
  recurrenceMonths: number | null;
  lastGeneratedAt: string | null;
  createdAt: string;
};

File: apps/api/Finly.Domain/Entities/Transaction.cs
using Finly.Domain.Common;
using Finly.Domain.Enums;

namespace Finly.Domain.Entities;

public class Transaction : BaseEntity
{
    public Guid FinancialProfileId { get; set; }

    public string Title { get; set; } = string.Empty;
    public decimal Amount { get; set; }
    public TransactionType Type { get; set; }
    public string Category { get; set; } = string.Empty;

    public TransactionKind TransactionKind { get; set; } = TransactionKind.Single;

    public DateOnly TransactionDate { get; set; }

    public Guid? SourceId { get; set; }

    public int? InstallmentIndex { get; set; }
    public int? InstallmentCount { get; set; }

    public bool IsRecurring { get; set; }
    public DateOnly? RecurrenceStartDate { get; set; }
    public DateOnly? RecurrenceEndDate { get; set; }
    public int? RecurrenceDay { get; set; }
    public int? RecurrenceMonths { get; set; }

    public FinancialProfile FinancialProfile { get; set; } = null!;
}
